src/api/schema/gen.rs

```diff
@@ -106,5 +106,6 @@ async fn main() {
     let res = schema.execute(INTROSPECTION_QUERY).await;
     let json = serde_json::to_string_pretty(&async_graphql::http::GQLResponse(res)).unwrap();
 
-    fs::write("graphql/schema.json", format!("{}\n", json)).expect("Couldn't save schema file");
+    fs::write("lib/api_client/graphql/schema.json", format!("{}\n", json))
```

Comment:
This is slightly overlapping-- the resulting `schema.json` is pushed into `lib/api_client`, rather than being top-level.

This is still generated by Vector proper (via `cargo run --bin graphql-schema`) since it's the API that's responsible for the schema; the client merely consumes it for code gen. So I think this is still the right place for it.

Reply:
👍 Yeah, that's a good idea. I'll make the switch.

Reply:
Hmm this might be an issue:

> The build script does not have access to the dependencies listed in the dependencies or dev-dependencies section (they’re not built yet!). Also, build dependencies are not available to the package itself unless also explicitly added in the [dependencies] table.

We require the inner `api` package to build the schema, so I'm not sure if this is feasible?

Reply:
We'll be able to pass it in the `build-dependencies`. They'll be two separate dependencies/builds, so it's not trivial whether we'd want to actually use it.

I wonder if there's a ready-made crate that would provide this introspection dumping for `build.rs` - seems like a pretty generic thing, so maybe we don't have to bother.

---

Might be nice to take the path as an argument rather than hard-coding it. Or, just somehow configuring it from the outside of the binary.

Reply:
I think the issue might be, the `api` code actually has to run for it to produce the resulting `schema.json`. All of the `#[derive(GraphQL)]` macros need expanding, and mant of the the result types overly internal metrics types, which necessitates Vector being compiled first.

So even if we were able to pull in `api` independently, wouldn't we also need the rest of Vector to be compiled alongside it? That's what I'm thinking.

I like the idea of this being an automatic part of the build process, but it doesn't seem like `build.rs` is designed for consuming the crate that it's pre-empting being built. Seems a bit chicken and egg.

Or am I thinking about this wrong?

Reply:
Ah, in this case we'd have to split the schema into a separate crate, both to avoid the chicken and egg problem and to reduce the compile times.

I wonder if it makes sense to move this `gen` binary to something like `vector api generate-schema` command, so anyone can use it, as it seems like all the data is in the binary either way.

Reply:
> Ah, in this case we'd have to split the schema into a separate crate, both to avoid the chicken and egg problem and to reduce the compile times.

I think that probably makes sense if/once topology and metrics are separate. The GraphQL API is always going to be tightly coupled to that, so if the rest of it lives in Vector, I think the API needs to, also. If we get to a point where that's pulled out, it'll unblock pulling out the API, too.

> I wonder if it makes sense to move this gen binary to something like vector api generate-schema command, so anyone can use it, as it seems like all the data is in the binary either way.

I like that idea. That could be useful for third-party tooling, where users are building their own clients around the API (of course, we could also have a hosted `schema.json` -- either internally, or in docs, or both.)

I'll kick that can down the road a bit, since it's not the immediate scope of this PR, but I think it's worth creating an issue for.